baseband/guppi/payload.py

```diff
@@ -0,0 +1,239 @@
+
+        # If we want to decode the entire dataset.
+        if item is () or item == slice(None):
+            data = decoder(self.words)
+            if self.complex_data:
+                data = data.view(self.dtype)
+            if self.time_ordered:
+                # Reshape to (nchan, nsample, npol); transpose to usual order.
+                return (data.reshape(self.sample_shape[1], -1,
+                                     self.sample_shape[0])
+                        .transpose(1, 2, 0))
+            else:
+                # Reshape to (nsample, nchan, npol); transpose to usual order.
+                return (data.reshape(-1, self.sample_shape[1],
+                                     self.sample_shape[0])
+                        .transpose(0, 2, 1))
+
+        words_slice, data_slice = self._item_to_slices(item)
+
+        if self.time_ordered:
+            # Reshape words so channels fall along first axis, then decode.
+            decoded_words = decoder(
+                self.words.reshape(self.sample_shape[1], -1)[:, words_slice])
+            return (decoded_words.view(self.dtype).T
+                    .reshape(-1, *self.sample_shape)[data_slice])
+        else:
+            # data_slice assumes (npol, nchan), so transpose before using it.
+            return (decoder(self.words[words_slice]).view(self.dtype)
+                    .reshape(-1, self.sample_shape[1], self.sample_shape[0])
+                    .transpose(0, 2, 1)[data_slice])
+
+    def __setitem__(self, item, data):
+        if item is () or item == slice(None):
+            words_slice = data_slice = slice(None)
+        else:
+            words_slice, data_slice = self._item_to_slices(item)
+
+        data = np.asanyarray(data)
+        # Check if the new data spans an entire word and is correctly shaped.
+        # If so, skip decoding.  If not, decode appropriate words and insert
+        # new data.
+        if not (data_slice == slice(None) and
+                data.shape[-len(self.sample_shape):] == self.sample_shape and
+                data.dtype.kind == self.dtype.kind):
+            decoder = self._decoders[self._coder]
+            if self.time_ordered:
+                decoded_words = decoder(np.ascontiguousarray(
+                    self.words.reshape(
+                        self.sample_shape[1], -1)[:, words_slice]))
+                current_data = (decoded_words.view(self.dtype)
+                                .T.reshape(-1, *self.sample_shape))
+            else:
+                current_data = (decoder(self.words[words_slice])
+                                .view(self.dtype)
+                                .reshape(-1, self.sample_shape[1],
+                                         self.sample_shape[0])
+                                .transpose(0, 2, 1))
+
+            current_data[data_slice] = data
+            data = current_data
+
+        if self.time_ordered:
+            data = data.reshape(-1, self.sample_shape.nchan).T
```

Comment:
Why is this reshaped to a 2-d rather than a 3-d array? I guess this question is for above too. I think it would be clearer to always work with the 3-d structure, even if strictly it makes no difference.

Reply:
As written, I think the code can be used even if `bps` is not 8 (according to http://seraproject.org/mw/images/b/b6/RawSlide7.jpeg, there's an integer number of bytes in each channel per frame), but it wouldn't be possible to separate the polarization samples without decoding first.  I also don't I agree that working in 3D always makes more sense, as I basically have to write out the reshapes and transposes each time I deal with this method.